BLSQ/iaso
# POLIO-1125 Multi Round Budget (Backend)

Implement Polio Multi Round Budget.

Related JIRA tickets : [POLIO-1125](https://bluesquare.atlassian.net/browse/POLIO-1125)

## Important: this PR cannot be merged "as is" because it will break the front-end

I have listed below the API changes that the front-end must implement [to make this happen](https://whimsical.com/multi-round-budget-FdiBzSUsuozi8aAnuvJWAP).

## Changes

### 1) Models

<img width="1515" alt="before_after" src="https://github.com/BLSQ/iaso/assets/281139/4fda481d-de50-4a6e-a3bc-fece81639d2c">

Notable changes on `Campaign` model:

1. copied budget related fields to `BudgetProcess`
    - these fields must be removed from `Campaign` after this feature is proven successful
1. grouped a bunch of deprecated fields at the end of the model
    - we should clean these fields as they add layers of indirection

### 2) Data migration

New data migration to move data between `Campaign` and `BudgetProcess`.

This should be thoroughly reviewed as well (as the `BudgetProcess` model fields).

### 3) Admin

- new admin UI for `BudgetProcess`

### 4) Tests

- moved tests from `plugins.polio.budget.test` to `plugins.polio.api.budget`
- renamed the test class `TeamAPITestCase` to `BudgetCampaignViewSetTestCase`
- renamed 1 letter long variables ([see original file](https://github.com/BLSQ/iaso/blob/f968dc97d2b9c4a387afdc664ed99d402d32a2d6/plugins/polio/budget/test.py))
- added tests for previously untested things:
    - `CampaignBudgetSerializer.get_timeline()`
    - `TransitionOverrideSerializer`
        - fixed a permission bug making the path unreachable for non admin users
    - workflow categories
    - etc.

Note that there could still be untested areas of the code.

### 5) Serializers

`CampaignBudgetSerializer` no longer inherits from `CampaignSerializer` as it is now related to the `BudgetProcess`.

As a result, I deleted the following fields from `CampaignBudgetSerializer`:

- `cvdpv2_notified_at`
- `ordering_fields` (all were linked to `Campaign`)
- `filterset_fields` (all were linked to `Campaign`)

### 6) API

List of API changes to make it easier for the front-end.

_Not that things can be modified according to front-end needs_.

#### `POST /api/polio/budget/` (create a `BudgetProcess` instance)

POST data:

```
data={
    "rounds": [round_1.id, round_2.id],
},
```

Return:

```
{
    'id': 5,                // BudgetProcess ID.
    'created_by': {
        'first_name': 'test',
        'last_name': 'test',
        'username': 'test'
    },
    'created_at': '2024-02-08T15:39:55.648934Z',
    'rounds': [8, 9]
}
```

#### `GET /api/polio/budget/` (get a list of `BudgetProcess` instances)

Return before:

```
{
    'results': [{
        'created_at': '2024-02-08T15:22:24.328053Z',
        'id': '43e53995-cc54-4f2a-985c-f81d2b6d67a3', // Campaign ID.
        'obr_name': 'test campaign',
        'country_name': None,
        'current_state': {
            'key': '-',
            'label': '-'
        },
        'budget_last_updated_at': None
    }]
}
```

Return after:

```
{
    'results': [{
        'created_at': '2024-02-08T15:12:38.937758Z',
        'id': 17,                                   // BudgetProcess ID.
        'obr_name': 'test campaign',
        'country_name': 'ANGOLA',
        'round_numbers': [1, 2],                    // List of **numbers** of linked rounds.
        'current_state': {
            'key': '-',
            'label': '-'
        },
        'updated_at': '2024-02-08T15:12:38.937774Z'
    }]
}
```

#### `GET /api/polio/budget/id/` (get a `BudgetProcess` instance)

Before:

```
GET /api/polio/budget/campaign_id/
```

After:

```
GET /api/polio/budget/budget_process_id/
```

#### `POST /api/polio/budget/transition_to/` (make a `BudgetProcess` transition to the next step)

POST data before:

```
data={
    "transition_key": "submit_budget",
    "campaign": campaign.id,                 // Campaign ID.
    "comment": "comment",
    "files": [],
}
```

POST data after:

```
data={
    "transition_key": "submit_budget",
    "budget_process": budget_process.id,    // BudgetProcess ID.
    "comment": "comment",
    "files": [],
}
```

#### `POST /api/polio/budget/override/` (make a `BudgetProcess` transition to an arbitrary step)

POST data before:

```
{
    "new_state_key": "rejected",
    "campaign": campaign.id,                 // Campaign ID.
    "comment": "override me",
    "files": [],
}
```

POST data after:

```
{
    "new_state_key": "rejected",
    "budget_process": budget_process.id,    // BudgetProcess ID.
    "comment": "override me",
    "files": [],
}
```

## How to test

See unit tests.



[POLIO-1125]: https://bluesquare.atlassian.net/browse/POLIO-1125?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
